core/src/Cuda/Kokkos_Cuda_BlockSize_Deduction.hpp

```diff
@@ -48,18 +48,134 @@
 #include <Kokkos_Macros.hpp>
 #ifdef KOKKOS_ENABLE_CUDA
 
-#include <iostream>
+#include <sstream>
 #include <Cuda/Kokkos_Cuda_Error.hpp>
 
 namespace Kokkos {
 namespace Impl {
 
+inline int cuda_max_active_blocks_per_sm(cudaDeviceProp const& properties,
+                                         cudaFuncAttributes const& attributes,
+                                         int block_size, size_t dynamic_shmem) {
+  // Limits due do registers/SM
+  int const regs_per_sm     = properties.regsPerMultiprocessor;
+  int const regs_per_thread = attributes.numRegs;
+  int const max_blocks_regs = regs_per_sm / (regs_per_thread * block_size);
+
+  // Limits due to shared memory/SM
+  size_t const shmem_per_sm    = properties.sharedMemPerMultiprocessor;
+  size_t const shmem_per_block = properties.sharedMemPerBlock;
+  size_t const static_shmem    = attributes.sharedSizeBytes;
+  // Temporarily commented since was not used previously in block size deduction
+  // size_t const dynamic_shmem_per_block =
+  // attributes.maxDynamicSharedSizeBytes;
+  size_t const total_shmem = static_shmem + dynamic_shmem;
+
+  int const max_blocks_shmem =
+      total_shmem > shmem_per_block
+          //      total_shmem > shmem_per_block || dynamic_shmem >
+          //      dynamic_shmem_per_block
+          ? 0
+          : (total_shmem > 0 ? (int)shmem_per_sm / total_shmem
+                             : max_blocks_regs);
+
+  // Overall occupancy in blocks
+  return std::min(max_blocks_regs, max_blocks_shmem);
+}
+
+template <typename UnaryFunction, typename LaunchBounds>
+inline int cuda_deduce_block_size(bool early_termination,
+                                  cudaDeviceProp const& properties,
+                                  cudaFuncAttributes const& attributes,
+                                  UnaryFunction block_size_to_dynamic_shmem,
+                                  LaunchBounds) {
+  // Limits
+  int const max_threads_per_sm = properties.maxThreadsPerMultiProcessor;
+  // unsure if I need to do that or if this is already accounted for in the
+  // functor attributes
+  int const max_threads_per_block =
+      std::min(LaunchBounds::maxTperB == 0 ? (int)properties.maxThreadsPerBlock
+                                           : (int)LaunchBounds::maxTperB,
+               attributes.maxThreadsPerBlock);
+  int const min_blocks_per_sm =
+      LaunchBounds::minBperSM == 0 ? 1 : LaunchBounds::minBperSM;
+
+#if CUDA_VERSION >= 11000
+  int const max_blocks_per_sm = properties.maxBlocksPerMultiProcessor;
+#else
+  int const max_blocks_per_sm = [&properties]() {
+    switch (properties.major) {
+      case 3: return 16;
+      case 5:
+      case 6: return 32;
+      case 7: {
+        int isTuring = properties.minor == 5;
+        return (isTuring) ? 16 : 32;
+      }
+      default:
+        throw_runtime_exception("Unknown device in cuda block size deduction");
+    }
+  }();
+#endif
```

Comment:
@crtrott My intent is to account for this in `cuda_max_active_blocks_per_sm()` but I did not want to change too much of the current logic at once.